### ISSUE TITLE
Flat Role Claim(s)

### ISSUE THREAD

**Issue by fforootd** [zitadel/zitadel#3997]
As an administrator I want to be able to create an action to set a custom claim ("groups" / "permissions") in the complement token flow with the users roles in the desired format in order to fulfill my 3rd party tools needs.

**Acceptance criteria**
- [x] I am able to access (read) the asserted claims
- [x] I am able to load the users roles through the action's `ctx`
- [x] I am able to load the user through the action's `ctx`
- [x] I have an example for mapping multi tenant roles to groups / permissions
- [ ] Updated oauth2-proxy QuickStart

----

**Additional Info**

While taking with @rollicks-c around how to use ZITADEL's role model with the oauth2-proxy project the question came up if it is feasible to reuse the existing groups claim. Currently we provide the claim as json which breaks the groups claim parsing because it awaits as list of strings.

A possible solution to this that I already did discuss with @livio-a is to allow customers to request the claim in a flattened manor. For example `["{orgId}:{role}", "{orgId}:{role}"]`

My 2 cents is that we should do this in order to encourage more people to convert from keycloak. Is see more cases there as well. For example ArgoCD also could accept it in that way.

**Comment by rollicks-c** [comment]
Hi Zitadel-Team.

Just to be sure - this issue is not covered by the resolved issue #3996, right?

**Comment by livio-a** [comment]
> Hi Zitadel-Team.
> 
> Just to be sure - this issue is not covered by the resolved issue #3996, right?

Hey @rollicks-c. No that one just ensured that we return the existing ones on the introspection endpoint as well.
But when we implement this issue, we'll certainly return them on the introspection as well.

**Comment by coelho** [comment]
Is there a work-around for this issue when using OAuth2 Proxy? From what I see there is no current work-around.

**Comment by rollicks-c** [comment]
Hi Zitadel-Team

Can you give us an update on the status of this issue?
We were informed that this will be implemented within 2 to 4 weeks.

As you know, we will need it very soon now. Thank you

**Comment by fforootd** [comment]
Hi @rollicks-c and @coelho 

In a few days (end of this month) we will release the option to defined custom claims with our actions feature.
This will allow you to create a small `js` snippet and use the roles from ZITADEL to add a claim to your tokens/userinfo with the flat roles (or in whatever format you like)

See the relating PR https://github.com/zitadel/zitadel/issues/4265

**Comment by rollicks-c** [comment]
Sweet. Looking forward to your next release. Thank you. :-)

**Comment by flyingfishflash** [comment]
I was trying to do something similar, take the roles created via UI and us JS via 'Complement Token' action to parse the existing roles of the user into a list or format them however I need them.

It's hard to tell from this issue, but are the user role grants (aka authorizations if I understand correctly) available in the context of an action in the Complement Token login flow?

Should I be using the http module to obtain them or maybe there's a different way I'm missing? 

Basically I'm trying to parse-reformat the existing user role grants into a usable format without having to recreate the role info as user metadata.

Thanks for this nice software btw!

**Comment by hifabienne** [comment]
I think this can be done with an action as soon as #4909 this is implemented. We will provide an example action when the other issue is done

**Comment by hifabienne** [comment]
@livio-a can you please rewrite this, so we know what we want to implement and can estimate it in the next refinement?

**Comment by bitfactory-sem-denbroeder** [comment]
Is there a current status / ETA for this issue? This is preventing adoption for zitadel in my organization.

**Comment by hifabienne** [comment]
Hei @bitfactory-sem-denbroeder 
The goal is to include this in the next sprint that starts next week. 
So it would be released mid march.

**Comment by adlerhurst** [comment]
@fforootd i don't understand what I need to update in the oauth 2 proxy guide. as well as i don't have access to the doc

**Comment by fforootd** [comment]
> @fforootd i don't understand what I need to update in the [oauth 2 proxy guide](https://docs-git-flat-role-claims-zitadel.vercel.app/docs/examples/identity-proxy/oauth2-proxy#configure-zitadel). as well as i don't have access to the doc

Why do you not have access to the docs? Its in the same repository as zitadel 😁 

You can start the oauth proxy with a oidc groups claim to test this:

![image](https://user-images.githubusercontent.com/9879976/223184246-1130b29d-9592-4ae5-91f5-d77a04267310.png)

https://oauth2-proxy.github.io/oauth2-proxy/docs/configuration/overview
